### PR TITLE
EW-FH New extension methods for properties with value list items (single/multiselect)

### PR DESCRIPTION
**Pull request by falk-huth-ewerk**
Background:
There are many helper methods for handling references to object types but to less for handling value lists and their items - especially in connection with those which where synchronised from external data sources - to have the same IDs on multiple vaults and other systems.

Rudimentary tests are already available but should be reviewed, Perhaps there are other ways to reach the goal more easily.

### CODE REVIEW

**Comment by falk-huth-ewerk** [comment]
@CraigHawker  Here again.

File: MFiles.VAF.Extensions/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItems.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            // Get the lookup elements and initialize result list
+            IEnumerable<Lookup> lookups = objVerEx.GetLookupsFromProperty(propDefId);
+            List<ValueListItem> listResults = new List<ValueListItem>();
+            VaultValueListItemOperations vliOps = objVerEx.Vault.ValueListItemOperations;
+
+            // Loop through the lookup elements
+            foreach (Lookup lookup in lookups)
+            {
+                listResults.Add(
+                    vliOps.GetValueListItemByID(propDef.ValueList, lookup.Item)
```
Comment (CraigHawker) [inline MFiles.VAF.Extensions/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItems.cs:80]: Have you tested what happens if the value list item has been deleted/destroyed?  Does it return null as the method XML states, or does it throw an exception?

File: MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItems.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        public void ThrowsIfNullObjVerEx()
+        {
+            ((Common.ObjVerEx) null).GetPropertyAsValueListItems((int) MFBuiltInPropertyDef.MFBuiltInPropertyDefWorkflow);
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentOutOfRangeException))]
+        public void ThrowsIfPropDefIdNegative()
+        {
+            ((Common.ObjVerEx) null).GetPropertyAsValueListItems((int) -1);
```
Comment (CraigHawker) [inline MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItems.cs:29]: This test does not pass.  The exception throws an ArgumentNullException - did you mean to use a valid ObjVerEx instance?

File: MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItems.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentException))]
+        public void ThrowsIfPropDoesNotExist()
+        {
+            // max int should not point to an existing property definition
+            ((Common.ObjVerEx) null).GetPropertyAsValueListItems(int.MaxValue);
```
Comment (CraigHawker) [inline MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItems.cs:37]: This test does not pass.  The exception throws an ArgumentNullException - did you mean to use a valid ObjVerEx instance?

File: MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItems.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentException))]
+        public void ThrowsIfPropIsNotBasedOnValueList()
+        {
+            ((Common.ObjVerEx) null).GetPropertyAsValueListItems((int) MFBuiltInPropertyDef.MFBuiltInPropertyDefDeleted);
```
Comment (CraigHawker) [inline MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItems.cs:44]: This test does not pass.  The exception throws an ArgumentNullException - did you mean to use a valid ObjVerEx instance?

File: MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItems.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentException))]
+        public void ThrowsIfPropIsNoMultiSelectLookup()
+        {
+            ((Common.ObjVerEx) null).GetPropertyAsValueListItems((int) MFBuiltInPropertyDef.MFBuiltInPropertyDefWorkflow);
```
Comment (CraigHawker) [inline MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItems.cs:51]: This test does not pass.  The exception throws an ArgumentNullException - did you mean to use a valid ObjVerEx instance?

File: MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItems.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        // TODO Think about further test methods + Question how lookups to object types were handled
```
Comment (CraigHawker) [inline MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItems.cs:54]: Some additional tests are needed here (could be done after merge, but must be done before release) to test valid scenarios (e.g. a valid property value type).

See examples on `GetPropertyAsSimpleTypeBase<T>` on how to mock the required structures.

File: MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItem.cs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        public void ThrowsIfNullObjVerEx()
+        {
+            ((Common.ObjVerEx) null).GetPropertyAsValueListItem((int) MFBuiltInPropertyDef.MFBuiltInPropertyDefWorkflow);
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentOutOfRangeException))]
+        public void ThrowsIfPropDefIdNegative()
+        {
+            ((Common.ObjVerEx) null).GetPropertyAsValueListItem((int) -1);
```
Comment (CraigHawker) [inline MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItem.cs:29]: This test does not pass.  The exception throws an ArgumentNullException - did you mean to use a valid ObjVerEx instance?

File: MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItem.cs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentException))]
+        public void ThrowsIfPropDoesNotExist()
+        {
+            // max int should not point to an existing property definition
+            ((Common.ObjVerEx) null).GetPropertyAsValueListItem(int.MaxValue);
```
Comment (CraigHawker) [inline MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItem.cs:37]: This test does not pass.  The exception throws an ArgumentNullException - did you mean to use a valid ObjVerEx instance?

File: MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItem.cs [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentException))]
+        public void ThrowsIfPropIsNotBasedOnValueList()
+        {
+            // max int should not point to an existing property definition
+            ((Common.ObjVerEx) null).GetPropertyAsValueListItem((int)MFBuiltInPropertyDef.MFBuiltInPropertyDefDeleted);
```
Comment (CraigHawker) [inline MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItem.cs:45]: This test does not pass.  The exception throws an ArgumentNullException - did you mean to use a valid ObjVerEx instance?

File: MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItem.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        }
+
+        // TODO Think about further test methods + Question how lookups to object types were handled
```
Comment (CraigHawker) [inline MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItem.cs:48]: Some additional tests are needed here (could be done after merge, but must be done before release) to test valid scenarios (e.g. a valid property value type).

See examples on GetPropertyAsSimpleTypeBase<T> on how to mock the required structures.

**Review by CraigHawker — CHANGES_REQUESTED** [review]
I have added some comments to the current submission.  In general I think these are very good methods to add, but at a minimum the tests need to pass before merge.

There should also be some tests to ensure that this executes correctly, not just testing error conditions.  I am happy to help with these once the above tests are resolved.

**Comment by falk-huth-ewerk** [comment]
@CraigHawker Tests run now correctly

Further tests are not possible as easy as you may think. vault.PropertyDefOperations returns empty PropertyDef set and you cannot add PropertyDef for testing. You can call AddPropertyDefAdmin but the new value doesn't get persisted because no real vault is connected. You have to implement a complete Vault mock for that.

**Comment by CraigHawker** [comment]
I have marked as resolved the comments related to the existing tests.  I have also done a merge from the primary branch as there was a conflict there.

File: MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItem.cs (hunk 10 above)
Comment (CraigHawker) [reply MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItem.cs]: Example unit tests included in commit 29f556895f81fb2232e1d7be75f406ee0f95f35a.

File: MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItems.cs (hunk 6 above)
Comment (CraigHawker) [reply MFiles.VAF.Extensions.Tests/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItems.cs]: Commit d60f7e500c76c15a7eda71cbbd77ecf2ef7b78a0 includes examples of unit tests for `GetPropertyAsValueListItems`.

**Comment by CraigHawker** [comment]
Unit tests added and green.  Will merge later today.

File: MFiles.VAF.Extensions/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItems.cs (hunk 1 above)
Comment (CraigHawker) [reply MFiles.VAF.Extensions/ExtensionMethods/ObjVerEx/GetPropertyAsValueListItems.cs]: Items that are deleted are returned (provided the user has sufficient privileges, which is the case for VAF code), but have their "Deleted" property set to true.
Items that are not found throw an exception, not return null.  Commit 0de244ade1cc07d1ecd6f4299303612b0b130174 deals with this and alters the XML to describe the behaviour.

**Pull request merged**